proposals/0672-declarative-xcode-projects.md

```diff
@@ -0,0 +1,92 @@
+
+## Detailed design
+
+To adopt Tuist in React Native, we need to follow some steps:
+
+1. Land https://github.com/facebook/react-native/pull/37952. This PR already contains a functioning project which passed all the iOS tests in CI.
+2. Update the CLI to install Tuist in the user machine. Tuist comes directly with an additional executable called `tuistenv`, which handles the environment and version for the user.
+3. Update the CLI and instruct it to swap the `HelloWorld` strings in the `template/ios/Package.swift` to the AppName chosen by the user.
+4. Update the CLI to run `tuist generate` to create the project
+5. Update the CLI to run `tuist generate` on `yarn ios` or `npm run ios` in case the `Project.swift` has changed or there is no `Xcodeproj` file.
+6. Update doctor to check that the user have Tuist installed
+7. Remove the old Xcodeproj file and set the right `.gitignore` settings.
+
+## Drawbacks
+
+- Prioritization: I think that the only reason NOT to do this is for prioritization and time. Implementing this system will take some time which we can devote to other, maybe more impactful, activities.
+- Impact: This work could simplify updates, but recently we have started pushing all the changes to the Xcodeproj into the Cocoapods scripts. This means that the users don't have to touch the project file at all, and Cocoapods takes care of modifying the project when needed. It is still useful to have a laid-out description of the project, especially for brownfield projects that have to mimic the same configurations of a template app.
+- Workflow changes: there is an extra step that user has to run when modifying the project: after every change in the `Project.swift`, users have to run `tuist generate`. This is a minor issue as we do not expect for the user to manually update the `Project.swift` file at all, and we can bake most of the Tuist invocations in the CLI.
+
+## Alternatives
+
+### [XCodeGen](https://github.com/yonaskolb/XcodeGen)
+
+This tool allows to describe the Xcodeproj using a YAML file.
+
+If we go for this approach, the drawbacks are:
+
+- a new language to teach to our iOS users
+- more verbose project description
+- no code completion.
+- Xcodegen is less flexible than Tuist as it is not executable code and it is not easily extensible.
+
+### [Xcake](https://github.com/igor-makarov/xcake/)
+
+This tool allows to describe a project in Ruby.
+
+If we go for this approach, the drawbacks are:
+
+- a new language to teach to our users
+- no type checking due to the dynamic nature of Ruby
+- one of the oldest tool: it has been unmaintained for a few years before some other OSS contributor took ownership of it. It could go unmaintained again.
+
+## Adoption strategy
+
+- Implement all the changes before 0.73.
```

Comment:
What does this mean for users who can't / don't want to migrate to Tuist in 0.73? Is this a breaking change for them?

Reply:
Yes, it will be. This is a bold decision, but for the better. 

We can alleviate it, if we won't, by releasing 0.73 with Tuist AND by keeping the Tuist generated Xcodeproj (so they don't have to use tuist from 0.73). But I will go all in with the tool at least from 0.74.

Reply:
Resolving this as I added this part in the RFC